packages/pwa-kit-dev/src/utils/logger.js

```diff
@@ -0,0 +1,200 @@
+     * @param {string} message - The log message.
+     * @param {LogOptions} [options={}] - Optional message details.
+     */
+    #printLog(message, options = {}) {
+        const {level} = options
+
+        if (!this.#shouldLog(level)) {
+            return
+        }
+
+        const formattedMessage = this.#formatLogMessage(message, options)
+
+        switch (level) {
+            case 'error':
+                console.error(formattedMessage)
+                break
+            case 'warn':
+                console.warn(formattedMessage)
+                break
+            case 'info':
+                console.info(formattedMessage)
+                break
+            case 'debug':
+                console.debug(formattedMessage)
+                break
+            default:
+                console.log(formattedMessage)
+                break
+        }
+    }
+
+    /**
+     * Logs a debug message.
+     * @param {string} message - The debug message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    debug(message, details = {}) {
+        this.#printLog(message, {level: 'debug', ...details})
+    }
+
+    /**
+     * Logs an info message.
+     * @param {string} message - The info message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    log(message, details = {}) {
+        this.#printLog(message, {level: 'log', ...details})
+    }
+
+    /**
+     * Logs an info message.
+     * @param {string} message - The info message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    info(message, details = {}) {
+        this.#printLog(message, {level: 'info', ...details})
+    }
+
+    /**
+     * Logs a warning message.
+     * @param {string} message - The warning message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    warn(message, details = {}) {
+        this.#printLog(message, {level: 'warn', ...details})
+    }
+
+    /**
+     * Logs an error message.
+     * @param {string} message - The error message.
+     * @param {LogDetails} [details={}] - Optional message details.
+     */
+    error(message, details = {}) {
+        this.#printLog(message, {level: 'error', ...details})
+    }
+}
+
+/**
+ * Create a logger instance for each package.
+ * @param {string} packageName - The name of the package where the logger is used.
+ * @param {Object} [clientConfig={}] - Configuration options specific to the client environment.
+ * @param {string} [clientConfig.logLevel=loggerConfig.client.logLevel] - The log level for the client environment.
+ * @param {string} [clientConfig.format=loggerConfig.client.format] - The log format for the client environment.
+ * @returns {PWAKITLogger} - An instance of PWAKITLogger configured for the specified package.
+ */
+const createLogger = (packageName, clientConfig = {}) => {
```

Comment:
I see this `clientConfig`.. I guess I was expecting the environment variable to be able to change the log level/format for _both_ server and client sides.

But on 2nd thought, it's perhaps best to have them separate. Probably customers would like to have more logging visible on the server side, and less on the client side. So they'll need to be able to configure the server and client-side logging separately 👍